cf-serverd/server_common.c

```diff
@@ -271,231 +252,170 @@ int MatchClasses(EvalContext *ctx, ServerConnectionState *conn)
-                /* We have a pattern covering this class - so are we allowed to activate it? */
-                if ((IsMatchItemIn(ap->accesslist, conn->ipaddr)) ||
-                    (IsRegexItemIn(ctx, ap->accesslist, conn->hostname)) ||
-                    (IsRegexItemIn(ctx, ap->accesslist, userid1)) ||
-                    (IsRegexItemIn(ctx, ap->accesslist, userid2)) ||
-                    (IsRegexItemIn(ctx, ap->accesslist, conn->username)))
-                {
-                    Log(LOG_LEVEL_VERBOSE, "Attempt to define role/class %s is permitted", RlistScalarValue(rp));
-                    permitted = true;
-                }
-                else
-                {
-                    Log(LOG_LEVEL_VERBOSE, "Attempt to define role/class %s is denied", RlistScalarValue(rp));
-                    RlistDestroy(defines);
-                    return false;
-                }
+                Log(LOG_LEVEL_DEBUG,
+                    "EXEC: authorized class: %s", class);
+
+                return true;
             }
         }
-
     }
 
-    if (permitted)
-    {
-        Log(LOG_LEVEL_VERBOSE, "Role activation allowed");
-    }
-    else
-    {
-        Log(LOG_LEVEL_VERBOSE, "Role activation disallowed - abort execution");
-    }
-
-    RlistDestroy(defines);
-    return permitted;
+    Log(LOG_LEVEL_INFO,
+        "EXEC: denied class: %s", class);
+    return false;
 }
 
-static int OptionFound(char *args, char *pos, char *word)
-/*
- * Returns true if the current position 'pos' in buffer
- * 'args' corresponds to the word 'word'.  Words are
- * separated by spaces.
- */
+void DoExec(const ServerConnectionState *conn, const char *args)
 {
-    size_t len;
+    char sendbuf[CF_BUFSIZE - CF_INBAND_OFFSET];
+
+    /* args came through a transaction packet, usual limits apply. */
+    assert(strlen(args) < CF_BUFSIZE);
 
-    if (pos < args)
+    if (CFRUNCOMMAND[0] == '\0')
     {
-        return false;
+        char *message = "EXEC denied: no cfruncommand defined";
+        Log(LOG_LEVEL_INFO, "%s", message);
+        SendTransaction(conn->conn_info, message, 0, CF_DONE);
+        return;
     }
 
-/* Single options do not have to have spaces between */
+    char cmdbuff[CF_EXPANDSIZE];
+    int ret = snprintf(cmdbuff, CF_BUFSIZE, "%s -I -D",
+                       CFRUNCOMMAND);
+    /* TODO what if cfruncommand is not cf-agent?
+       These arguments don't apply then. */
 
-    if ((strlen(word) == 2) && (strncmp(pos, word, 2) == 0))
+    if (ret <= 0 || ret >= CF_BUFSIZE)
     {
-        return true;
+        /* impossible? */
+        return;
     }
 
-    len = strlen(word);
+    /* Pointer to append list of classes. */
+    char  *authorized_classes     = &cmdbuff[ret];
+    size_t authorized_classes_len = 0;
 
-    if (strncmp(pos, word, len) != 0)
-    {
-        return false;
-    }
+    StringAppendDelimited(authorized_classes, &authorized_classes_len,
+                          CF_BUFSIZE, "cfruncommand", ',');
```

Comment:
``CF_BUFSIZE`` happens to be half of ``CF_EXPANDSIZE``, so your content will always fit. However, perhaps it is better to be explicit here, and declare ``cmdbuff[CF_BUFSIZE*2]``? That would also protect you in case either constant is changed in the future and the relationship is not half/double anymore.